src/ucs/async/thread.h

```diff
@@ -9,13 +9,52 @@
 
 #include <ucs/type/spinlock.h>
 #include <ucs/sys/checker.h>
+#include <ucs/debug/assert.h>
+
+
+typedef struct ucs_async_thread_mutex {
+    pthread_mutex_t lock;
+    pthread_t       owner;
+    unsigned        count;
+} ucs_async_thread_mutext_t;
 
 
 typedef struct ucs_async_thread_context {
     union {
-        ucs_recursive_spinlock_t spinlock;
-        pthread_mutex_t          mutex;
+        ucs_recursive_spinlock_t  spinlock;
+        ucs_async_thread_mutext_t mutex;
     };
 } ucs_async_thread_context_t;
 
+
+static UCS_F_ALWAYS_INLINE int
+ucs_recursive_mutex_is_locked(const ucs_async_thread_mutext_t *mutex)
+{
+    return mutex->owner == pthread_self();
+}
+
+
+static UCS_F_ALWAYS_INLINE void
+ucs_recursive_mutex_block(ucs_async_thread_mutext_t *mutex)
+{
+    (void)pthread_mutex_lock(&mutex->lock);
+    ucs_assert((mutex->owner == UCS_THREAD_OWNER_NULL) ||
+               /* can be locked again by the same thread, because it is
+                * recursive mutex */
+               (ucs_recursive_mutex_is_locked(mutex) &&
+                (mutex->count > 0)));
+    mutex->count++;
+    mutex->owner = pthread_self();
```

Comment:
may be no need to set, as it is supposed to be self already?

Reply:
done